File: py/tests/test_elastix.py
"""Elastix transforms, checked against Elastix's own `transformix` output.

The ground truth in `transformix_golden.csv` was produced by the `transformix` binary itself, so
these are not self-consistency checks: they pin us to Elastix's behaviour, including its habit of
returning points outside the control-point grid **unchanged** rather than failing them.
"""

import pickle

import numpy as np
import pytest

import navis_fastcore as fastcore

# Measured margin is ~5e-7 (transformix's own print precision, six decimals). If a change lands
# at 1e-5, something is subtly wrong -- do not loosen this.
ATOL = 1e-4

CASES = ["affine", "translation", "euler", "euler_zyx", "similarity", "bspline", "add"]


@pytest.fixture(scope="module")
def bspline(elastix_dir):
    """The B-spline fixture, composed over an affine."""
    return fastcore.ElastixTransform(elastix_dir / "bspline.txt")


@pytest.fixture(scope="module")
def in_grid(elastix_dir, bspline):
    """Does a *source* point reach the spline's valid region (i.e. after the initial affine)?"""
    affine = fastcore.ElastixTransform(elastix_dir / "affine.txt")
    size = bspline.grid_size[0]
    origin = bspline.grid_origin[0]
    spacing = bspline.grid_spacing[0]

    def check(points):
        u = (affine.xform(np.atleast_2d(points)) - origin) / spacing
        return np.all((u >= 1.0) & (u < size - 2), axis=1)

    return check


# --- Against transformix ---------------------------------------------------------------


@pytest.mark.parametrize("case", CASES)
def test_matches_transformix(case, elastix_dir, elastix_golden):
    xf = fastcore.ElastixTransform(elastix_dir / f"{case}.txt")
    got = xf.xform(elastix_golden["input"])
    np.testing.assert_allclose(got, elastix_golden[case], atol=ATOL)


def test_compute_zyx_changes_the_rotation_order(elastix_dir, elastix_golden):
    """`euler.txt` and `euler_zyx.txt` carry identical angles and differ only in `ComputeZYX`.

    The default (false) is Rz*Rx*Ry, not the Rz*Ry*Rx its name suggests.
    """
    a = fastcore.ElastixTransform(elastix_dir / "euler.txt").xform(elastix_golden["input"])
    b = fastcore.ElastixTransform(elastix_dir / "euler_zyx.txt").xform(elastix_golden["input"])
    assert np.abs(a - b).max() > 1.0


# --- The grid boundary -----------------------------------------------------------------


def test_out_of_grid_points_are_returned_unchanged(bspline, elastix_dir):
    """Elastix's rule, and the exact opposite of CMTK's, which reports such points as FAILED."""
    far = np.array([[-5000.0, -5000.0, -5000.0]])

    # ...but the affine still applies: it is a separate step of the chain.
    affine = fastcore.ElastixTransform(elastix_dir / "affine.txt")
    np.testing.assert_allclose(bspline.xform(far), affine.xform(far), atol=1e-9)

    out = bspline.xform(far, out_of_bounds="nan")
    assert np.isnan(out).all()


def test_out_of_bounds_only_affects_out_of_grid_points(bspline, elastix_golden):
    """Points inside the grid must be unaffected by the flag."""
    got_id = bspline.xform(elastix_golden["input"])
    got_nan = bspline.xform(elastix_golden["input"], out_of_bounds="nan")

    inside = ~np.isnan(got_nan).any(axis=1)
    assert inside.sum() > 0 and (~inside).sum() > 0, "fixture must span the boundary"
    np.testing.assert_allclose(got_id[inside], got_nan[inside], atol=1e-12)


def test_bad_out_of_bounds_raises(bspline):
    with pytest.raises(ValueError, match="out_of_bounds"):
        bspline.xform(np.zeros((1, 3)), out_of_bounds="bogus")


# --- The inverse -----------------------------------------------------------------------


def test_inverse_is_forward_consistent(bspline, elastix_golden):
    """The guarantee: what we hand back really is a preimage."""
    fwd = bspline.xform(elastix_golden["input"])
    back = bspline.xform_inv(fwd)
    assert np.isfinite(back).all()
    np.testing.assert_allclose(bspline.xform(back), fwd, atol=1e-6)


def test_inverse_round_trips_for_sources_inside_the_grid(bspline, elastix_golden, in_grid):
    """A source point inside the grid has a *unique* warp preimage, so we recover it exactly.

    (The fixture is injective: det(I + J) > 0 throughout. A folded registration would not be.)
    """
    src = elastix_golden["input"]
    inside = in_grid(src)
    assert inside.sum() > 20, "fixture must have points inside the grid"

    back = bspline.xform_inv(bspline.xform(src))
    np.testing.assert_allclose(back[inside], src[inside], atol=1e-6)


def test_a_target_can_have_two_preimages(bspline, elastix_golden, in_grid):
    """The forward map is `p + disp(p)` inside the grid and `p` outside -- discontinuous, and so
    **not injective**. A target near the boundary can have a preimage on each branch.

    We prefer the warp branch, so for such a point the inverse returns a *different* point than
    you started from. That is not a defect and cannot be fixed; it is what a 2-to-1 map means.
    What we guarantee is that the answer is a genuine preimage -- which this asserts.
    """
    src = elastix_golden["input"]
    fwd = bspline.xform(src)
    back = bspline.xform_inv(fwd)

    off = np.abs(back - src).max(axis=1) > 1e-6
    assert off.sum() == 1, "fixture should carry exactly one such point"
    assert not in_grid(src)[off].any(), "...and its source must be outside the grid"

    # ...and what we returned really does map to the target.
    np.testing.assert_allclose(bspline.xform(back[off]), fwd[off], atol=1e-6)


def test_seed_iter_does_not_change_a_converged_answer(bspline, elastix_golden):
    """On a grid this small the seed is not *needed* -- see the real-file test below, where it is.

    This only pins that turning it off does not corrupt an answer the solver already gets right.
    """
    fwd = bspline.xform(elastix_golden["input"])
    np.testing.assert_allclose(
        bspline.xform_inv(fwd, seed_iter=0), bspline.xform_inv(fwd), atol=1e-6
    )


def test_add_chains_cannot_be_inverted(elastix_dir):
    """`T(x) = T0(x) + T1(x) - x` does not decompose into invertible hops. We refuse."""
    xf = fastcore.ElastixTransform(elastix_dir / "add.txt")
    with pytest.raises(ValueError, match="cannot be inverted"):
        xf.xform_inv(np.zeros((1, 3)))
    # ...but it still transforms forwards
    assert np.isfinite(xf.xform(np.zeros((1, 3)))).all()


def test_initial_guess(bspline, elastix_golden):
    """Handed the exact answer as a guess, the solver lands on it -- including for the ambiguous
    point, where the guess is what breaks the tie."""
    src = elastix_golden["input"]
    fwd = bspline.xform(src)
    back = bspline.xform_inv(fwd, initial_guess=src)
    np.testing.assert_allclose(back, src, atol=1e-6)

    with pytest.raises(ValueError, match="one point per input point"):
        bspline.xform_inv(fwd, initial_guess=src[:2])


# --- Chaining --------------------------------------------------------------------------


def test_initial_transform_is_resolved_from_the_files_own_directory(bspline):
    """`bspline.txt` names `affine.txt` by a bare relative filename."""
    assert bspline.kinds == [["linear", "bspline"]]
    assert len(bspline) == 1
    assert bspline.affine.shape == (4, 4)
    assert bspline.grid_size.tolist() == [[10, 10, 8]]


def test_chain_equals_manual_double_application(elastix_dir, elastix_golden):
    path = elastix_dir / "affine.txt"
    one = fastcore.ElastixTransform(path)
    two = fastcore.ElastixTransform([path, path])
    np.testing.assert_allclose(
        two.xform(elastix_golden["input"]),
        one.xform(one.xform(elastix_golden["input"])),
        atol=1e-9,
    )


def test_invert_flag_reverses_a_hop(elastix_dir, elastix_golden):
    """The `invert` flag must agree exactly with `xform_inv`."""
    path = elastix_dir / "bspline.txt"
    xf = fastcore.ElastixTransform(path)
    fwd = xf.xform(elastix_golden["input"])
    np.testing.assert_array_equal(
        fastcore.ElastixTransform(path, invert=True).xform(fwd), xf.xform_inv(fwd)
    )


def test_invert_length_mismatch_raises(elastix_dir):
    with pytest.raises(ValueError, match="one flag per transform"):
        fastcore.ElastixTransform([elastix_dir / "affine.txt"], invert=[True, False])


# --- Input handling --------------------------------------------------------------------


def test_single_point_in_single_point_out(bspline):
    got = bspline.xform([30.0, 25.0, 20.0])
    assert got.shape == (3,)
    np.testing.assert_allclose(got, bspline.xform(np.array([[30.0, 25.0, 20.0]]))[0])


def test_int_input_is_coerced(bspline):
    np.testing.assert_allclose(
        bspline.xform(np.array([[30, 25, 20]])),
        bspline.xform(np.array([[30.0, 25.0, 20.0]])),
    )


def test_bad_shape_raises(bspline):
    with pytest.raises(ValueError, match=r"\(N, 3\)"):
        bspline.xform(np.zeros((4, 2)))


def test_missing_file_raises(elastix_dir):
    with pytest.raises(ValueError, match="could not read"):
        fastcore.ElastixTransform(elastix_dir / "nope.txt")


def test_a_file_that_is_not_a_transform_raises(elastix_dir, tmp_path):
    """`navis-flybrains` ships a `template_to_BANC.txt` holding nothing but a filename."""
    p = tmp_path / "pointer.txt"
    p.write_text("3_elastix_Bspline_fine.txt\n")
    with pytest.raises(ValueError, match="not an elastix transform"):
        fastcore.ElastixTransform(p)


# --- Plumbing --------------------------------------------------------------------------


def test_pickle_round_trip(bspline, elastix_golden):
    """Pickling must re-load from disk, not ship the coefficients (BANC's file is 56 MB)."""
    blob = pickle.dumps(bspline)
    assert len(blob) < 1000, f"pickle is {len(blob)} bytes -- is it carrying the grid?"

    revived = pickle.loads(blob)
    np.testing.assert_allclose(
        revived.xform(elastix_golden["input"]), bspline.xform(elastix_golden["input"])
    )


def test_n_cores_does_not_change_the_answer(bspline, elastix_golden):
    fwd = bspline.xform(elastix_golden["input"])
    np.testing.assert_array_equal(
        bspline.xform_inv(fwd, n_cores=1), bspline.xform_inv(fwd)
    )


def test_repr(bspline):
    assert "ElastixTransform" in repr(bspline)
    assert "linear+bspline" in repr(bspline)


def test_load_elastix_transform_factory(elastix_dir):
    xf = fastcore.load_elastix_transform(elastix_dir / "affine.txt")
    assert isinstance(xf, fastcore.ElastixTransform)


def test_large_batch_is_finite(bspline):
    rng = np.random.default_rng(0)
    pts = rng.uniform(0, 100, (100_000, 3))
    assert np.isfinite(bspline.xform(pts)).all()

File: R/nat.fastcore/tests/testthat/test-elastix.R
# Elastix transforms, checked against Elastix's own `transformix` output.
#
# The ground truth in `transformix_golden.csv` was produced by the `transformix` binary itself, so
# these are not self-consistency checks: they pin us to Elastix's behaviour, including its habit of
# returning points outside the control-point grid **unchanged** rather than failing them (which is
# the exact opposite of CMTK, hence the contrast with `test-cmtk.R`).

ex <- function(f) system.file("extdata", "elastix", f, package = "nat.fastcore")

golden <- local({
  g <- utils::read.csv(ex("transformix_golden.csv"))
  split(as.matrix(g[, c("x", "y", "z")]), g$case)[unique(g$case)] |>
    lapply(matrix, ncol = 3)
})

# Measured margin is ~5e-7 (transformix's own print precision, six decimals). If a change lands
# at 1e-5, something is subtly wrong -- do not loosen this.
ATOL <- 1e-4

pts <- golden$input
bspline <- elastix_read(ex("bspline.txt"))

test_that("every transform type matches transformix", {
  for (case in c("affine", "translation", "euler", "euler_zyx", "similarity", "bspline", "add")) {
    xf <- elastix_read(ex(paste0(case, ".txt")))
    expect_equal(elastix_xform(xf, pts), golden[[case]],
                 tolerance = ATOL, ignore_attr = TRUE, info = case)
  }
})

test_that("ComputeZYX selects a different rotation order", {
  # Identical angles; the files differ only in `ComputeZYX`. The default (false) is Rz*Rx*Ry,
  # not the Rz*Ry*Rx its name suggests.
  a <- elastix_xform(elastix_read(ex("euler.txt")), pts)
  b <- elastix_xform(elastix_read(ex("euler_zyx.txt")), pts)
  expect_gt(max(abs(a - b)), 1)
})

test_that("points outside the grid come back unchanged, not NaN", {
  far <- rbind(c(-5000, -5000, -5000))

  # ...but the affine still applies -- it is a separate step of the chain
  affine <- elastix_read(ex("affine.txt"))
  expect_equal(elastix_xform(bspline, far), elastix_xform(affine, far), tolerance = 1e-9)

  # ...unless you ask to see the boundary
  expect_true(all(is.nan(elastix_xform(bspline, far, out_of_bounds = "nan"))))
})

test_that("the inverse is forward-consistent", {
  # The guarantee: whatever comes back really is a preimage. (Round-trip identity is NOT
  # guaranteed -- the forward map is not injective. See `elastix_xform_inv`.)
  fwd <- elastix_xform(bspline, pts)
  back <- elastix_xform_inv(bspline, fwd)
  expect_true(all(is.finite(back)))
  expect_equal(elastix_xform(bspline, back), fwd, tolerance = 1e-6)
})

test_that("an initial guess pins down the answer", {
  fwd <- elastix_xform(bspline, pts)
  expect_equal(elastix_xform_inv(bspline, fwd, initial_guess = pts), pts, tolerance = 1e-6)
})

test_that("Add chains cannot be inverted", {
  # T(x) = T0(x) + T1(x) - x does not decompose into invertible hops. We refuse.
  add <- elastix_read(ex("add.txt"))
  expect_error(elastix_xform_inv(add, pts), "cannot be inverted")
  expect_true(all(is.finite(elastix_xform(add, pts))))  # ...but it still goes forwards
  expect_true(bspline$ptr$invertible())
})

test_that("chains and the invert flag", {
  path <- ex("affine.txt")
  one <- elastix_read(path)
  two <- elastix_read(c(path, path))
  expect_length(elastix_kinds(two), 2)
  expect_equal(elastix_xform(two, pts), elastix_xform(one, elastix_xform(one, pts)),
               tolerance = 1e-9)

  # traversing a transform backwards agrees exactly with inverting it -- from the same parse
  fwd <- elastix_xform(bspline, pts)
  expect_equal(elastix_xform(bspline, fwd, invert = TRUE), elastix_xform_inv(bspline, fwd))
})

test_that("the initial transform resolves from the file's own directory", {
  # `bspline.txt` names `affine.txt` by a bare relative filename.
  expect_equal(elastix_kinds(bspline), "linear+bspline")
  expect_equal(dim(elastix_affine(bspline)), c(4L, 4L))
  expect_equal(as.vector(elastix_grid_size(bspline)), c(10, 10, 8))
  expect_equal(as.vector(elastix_grid_spacing(bspline)), c(14, 14, 13))
})

test_that("input coercion", {
  expect_equal(elastix_xform(bspline, c(30, 25, 20)), elastix_xform(bspline, rbind(c(30, 25, 20))))
  expect_equal(elastix_xform(bspline, as.data.frame(pts)), elastix_xform(bspline, pts))
})

test_that("n_cores does not change the answer", {
  expect_equal(elastix_xform(bspline, pts, n_cores = 1), elastix_xform(bspline, pts))
})

test_that("errors are informative", {
  expect_error(elastix_read("/nonexistent/nope.txt"), "no Elastix transform")
  expect_error(elastix_xform(bspline, pts, invert = c(TRUE, FALSE)), "one flag per transform")
  expect_error(elastix_xform(bspline, matrix(1, 2, 2)), "\\(N, 3\\)")
  expect_error(elastix_xform(bspline, pts, out_of_bounds = "bogus"))
  expect_error(elastix_xform(42, pts), "elastix_transform")
  expect_error(elastix_xform_inv(bspline, pts, initial_guess = pts[1:2, ]),
               "one point per input point")
  expect_output(print(bspline), "elastix_transform")
})

test_that("the header-only probe agrees with a full parse", {
  # That agreement is the probe's whole contract: it reads the same chain and skips only the
  # coefficients, so it must never give a different answer.
  for (name in c("affine.txt", "euler.txt", "similarity.txt", "bspline.txt", "add.txt")) {
    probed <- elastix_probe_invertible(ex(name))
    parsed <- elastix_read(ex(name))$ptr$invertible()
    expect_equal(probed, parsed, info = name)
  }

  expect_true(elastix_probe_invertible(ex("bspline.txt")))
  expect_false(elastix_probe_invertible(ex("add.txt")))
  expect_error(elastix_probe_invertible("/nonexistent/nope.txt"), "no Elastix transform")
})

test_that("a stale absolute initial-transform path falls back to its basename", {
  # Elastix records the path as it was on the author's machine. `copy_files` in navis is
  # precisely this workaround, in disguise.
  d <- tempfile()
  dir.create(d)
  on.exit(unlink(d, recursive = TRUE))
  writeLines(
    c('(Transform "TranslationTransform")', "(TransformParameters 1 2 3)"),
    file.path(d, "parent.txt")
  )
  writeLines(
    c(
      '(Transform "TranslationTransform")', "(TransformParameters 10 20 30)",
      '(InitialTransformParametersFileName "/nowhere/on/this/machine/parent.txt")'
    ),
    file.path(d, "child.txt")
  )

  xf <- elastix_read(file.path(d, "child.txt"))
  expect_equal(elastix_kinds(xf), "linear+linear")
  expect_equal(elastix_xform(xf, rbind(c(0, 0, 0))), rbind(c(11, 22, 33)))
  expect_true(elastix_probe_invertible(file.path(d, "child.txt")))
})
